File: runa/compiler/lexer.py
"""
Runa Lexer

Tokenizes Runa source code with natural language syntax support.
Handles multi-word tokens, indentation-based scoping, and natural language operators.

Following the Runa Formal Grammar Specifications exactly.
"""

import re
from typing import List, Iterator, Optional, Tuple
from .tokens import Token, TokenType, KEYWORDS, MULTI_WORD_TOKENS

class LexerError(Exception):
    """Exception raised for lexical analysis errors."""
    def __init__(self, message: str, line: int, column: int):
        self.message = message
        self.line = line
        self.column = column
        super().__init__(f"Lexer error at line {line}, column {column}: {message}")

class RunaLexer:
    """
    Lexer for the Runa natural language programming syntax.
    
    Handles:
    - Natural language keywords and multi-word tokens
    - Indentation-based scoping with INDENT/DEDENT tokens
    - String literals with escape sequences
    - Numeric literals (integers and floats)
    - Identifiers following naming conventions
    - Comments using "Note:" syntax
    """
    
    def __init__(self, source: str):
        self.source = source
        self.position = 0
        self.line = 1
        self.column = 1
        self.indent_stack = [0]  # Stack to track indentation levels
        self.tokens: List[Token] = []
        
        # Regex patterns for token recognition
        self.patterns = {
            'whitespace': re.compile(r'[ \t]+'),
            'newline': re.compile(r'\r?\n'),
            'comment': re.compile(r'Note:.*'),
            'string': re.compile(r'"([^"\\]|\\.)*"'),
            'float': re.compile(r'\d+\.\d+'),
            'integer': re.compile(r'\d+'),
            'identifier': re.compile(r'[a-zA-Z_][a-zA-Z0-9_]*'),
            'punctuation': re.compile(r'[:,;.?!()[\]{}|]'),
        }
    
    def current_char(self) -> Optional[str]:
        """Get the current character, or None if at end."""
        if self.position >= len(self.source):
            return None
        return self.source[self.position]
    
    def peek_char(self, offset: int = 1) -> Optional[str]:
        """Peek at a character ahead by offset positions."""
        pos = self.position + offset
        if pos >= len(self.source):
            return None
        return self.source[pos]
    
    def advance(self) -> None:
        """Advance to the next character."""
        if self.position < len(self.source) and self.source[self.position] == '\n':
            self.line += 1
            self.column = 1
        else:
            self.column += 1
        self.position += 1
    
    def peek_ahead(self, count: int) -> str:
        """Peek ahead count characters."""
        end_pos = min(self.position + count, len(self.source))
        return self.source[self.position:end_pos]
    
    def match_multi_word_token(self) -> Optional[Tuple[TokenType, int]]:
        """Try to match multi-word tokens like 'is greater than'."""
        # Try longest matches first
        for phrase in sorted(MULTI_WORD_TOKENS.keys(), key=len, reverse=True):
            if self.source[self.position:self.position+len(phrase)].lower() == phrase.lower():
                # Check that it's a complete word boundary
                next_pos = self.position + len(phrase)
                if (next_pos >= len(self.source) or 
                    self.source[next_pos] in ' \t\n\r:(),;.[]{}'):
                    token_type = MULTI_WORD_TOKENS[phrase]
                    if isinstance(token_type, tuple):
                        # Return the first token type for compound tokens
                        return token_type[0], len(phrase)
                    return token_type, len(phrase)
        return None
    
    def tokenize_string(self) -> Token:
        """Tokenize a string literal."""
        start_line, start_column = self.line, self.column
        quote_char = self.current_char()
        self.advance()  # Skip opening quote
        
        value = ""
        while self.current_char() and self.current_char() != quote_char:
            if self.current_char() == '\\':
                self.advance()
                # Handle escape sequences
                escape_char = self.current_char()
                if escape_char == 'n':
                    value += '\n'
                elif escape_char == 't':
                    value += '\t'
                elif escape_char == 'r':
                    value += '\r'
                elif escape_char == '\\':
                    value += '\\'
                elif escape_char == '"':
                    value += '"'
                elif escape_char == "'":
                    value += "'"
                else:
                    value += escape_char or ''
                self.advance()
            else:
                value += self.current_char()
                self.advance()
        
        if not self.current_char():
            raise LexerError("Unterminated string literal", start_line, start_column)
        
        self.advance()  # Skip closing quote
        return Token(TokenType.STRING, value, start_line, start_column)
    
    def tokenize_number(self) -> Token:
        """Tokenize a numeric literal (integer or float)."""
        start_line, start_column = self.line, self.column
        
        # Check for float first (contains decimal point)
        float_match = self.patterns['float'].match(self.source, self.position)
        if float_match:
            value = float_match.group()
            for _ in range(len(value)):
                self.advance()
            return Token(TokenType.FLOAT, float(value), start_line, start_column)
        
        # Otherwise, it's an integer
        int_match = self.patterns['integer'].match(self.source, self.position)
        if int_match:
            value = int_match.group()
            for _ in range(len(value)):
                self.advance()
            return Token(TokenType.INTEGER, int(value), start_line, start_column)
        
        raise LexerError("Invalid number format", start_line, start_column)
    
    def tokenize_identifier_or_keyword(self) -> Token:
        """Tokenize an identifier or keyword."""
        start_line, start_column = self.line, self.column
        
        # First try multi-word tokens
        multi_word = self.match_multi_word_token()
        if multi_word:
            token_type, length = multi_word
            value = self.source[self.position:self.position + length]
            for _ in range(length):
                self.advance()
            return Token(token_type, value, start_line, start_column)
        
        # Check for specific multi-word identifier patterns
        multi_word_patterns = [
            "user name", "user age", "tax rate", "total price", 
            "shopping cart", "file name", "last name", "first name"
        ]
        
        for pattern in multi_word_patterns:
            if (self.source[self.position:self.position+len(pattern)] == pattern and
                (self.position + len(pattern) >= len(self.source) or 
                 self.source[self.position + len(pattern)] in ' \t\n\r:(),;.[]{}beisgrlto')):
                # It's the pattern followed by a delimiter or keyword
                for _ in range(len(pattern)):
                    self.advance()
                return Token(TokenType.IDENTIFIER, pattern, start_line, start_column)
        
        # Single word identifier/keyword
        identifier_match = self.patterns['identifier'].match(self.source, self.position)
        if identifier_match:
            value = identifier_match.group()
            for _ in range(len(value)):
                self.advance()
            
            # Check if it's a keyword
            if value in KEYWORDS:
                token_type = KEYWORDS[value]
                # Handle boolean literals specially
                if token_type == TokenType.BOOLEAN:
                    bool_value = value.lower() == 'true'
                    return Token(TokenType.BOOLEAN, bool_value, start_line, start_column)
                return Token(token_type, value, start_line, start_column)
            
            # It's an identifier  
            return Token(TokenType.IDENTIFIER, value, start_line, start_column)
        
        raise LexerError(f"Invalid identifier: {self.current_char()}", start_line, start_column)
    

    
    def handle_indentation(self, indent_level: int) -> List[Token]:
        """Handle indentation changes and generate INDENT/DEDENT tokens."""
        tokens = []
        current_indent = self.indent_stack[-1]
        
        if indent_level > current_indent:
            # Increase in indentation
            self.indent_stack.append(indent_level)
            tokens.append(Token(TokenType.INDENT, indent_level, self.line, 1))
        elif indent_level < current_indent:
            # Decrease in indentation - may need multiple DEDENTs
            while self.indent_stack and self.indent_stack[-1] > indent_level:
                self.indent_stack.pop()
                tokens.append(Token(TokenType.DEDENT, indent_level, self.line, 1))
            
            if not self.indent_stack or self.indent_stack[-1] != indent_level:
                raise LexerError("Inconsistent indentation", self.line, 1)
        
        return tokens
    
    def tokenize(self) -> List[Token]:
        """
        Tokenize the entire source code.
        
        Returns:
            List of tokens representing the source code
            
        Raises:
            LexerError: If lexical analysis fails
        """
        self.tokens = []
        
        while self.position < len(self.source):
            # Handle newlines and indentation
            if self.current_char() == '\n':
                # Add newline token
                self.tokens.append(Token(TokenType.NEWLINE, '\n', self.line, self.column))
                self.advance()
                
                # Check indentation on next line
                if self.position < len(self.source):
                    indent_level = 0
                    while (self.position < len(self.source) and 
                           self.current_char() in ' \t'):
                        if self.current_char() == ' ':
                            indent_level += 1
                        elif self.current_char() == '\t':
                            indent_level += 4  # Tab = 4 spaces
                        self.advance()
                    
                    # Skip empty lines
                    if (self.position < len(self.source) and 
                        self.current_char() not in '\n\r'):
                        # Handle indentation changes
                        indent_tokens = self.handle_indentation(indent_level)
                        self.tokens.extend(indent_tokens)
                continue
            
            # Skip whitespace (but not newlines)
            if self.current_char() in ' \t':
                self.advance()
                continue
            
            # Handle comments
            if self.source[self.position:].startswith('Note:'):
                comment_match = self.patterns['comment'].match(self.source, self.position)
                if comment_match:
                    # Skip the entire comment
                    for _ in range(len(comment_match.group())):
                        self.advance()
                continue
            
            # Handle string literals
            if self.current_char() in '"\'':
                self.tokens.append(self.tokenize_string())
                continue
            
            # Handle numeric literals
            if self.current_char().isdigit():
                self.tokens.append(self.tokenize_number())
                continue
            
            # Handle punctuation
            if self.current_char() in '(),;.?![]{}:':
                char = self.current_char()
                token_type = {
                    '(': TokenType.LPAREN,
                    ')': TokenType.RPAREN,
                    '[': TokenType.LBRACKET,
                    ']': TokenType.RBRACKET,
                    '{': TokenType.LBRACE,
                    '}': TokenType.RBRACE,
                    ':': TokenType.COLON,
                    ',': TokenType.COMMA,
                    '.': TokenType.DOT,
                    ';': TokenType.SEMICOLON,
                    '?': TokenType.QUESTION,
                    '!': TokenType.EXCLAMATION,
                }[char]
                
                self.tokens.append(Token(token_type, char, self.line, self.column))
                self.advance()
                continue
            
            # Handle pipe for pattern matching
            if self.current_char() == '|':
                self.tokens.append(Token(TokenType.PIPE, '|', self.line, self.column))
                self.advance()
                continue
            
            # Handle underscore wildcard
            if self.current_char() == '_':
                # Check if it's standalone wildcard or part of identifier
                if (self.peek_char() is None or 
                    not self.peek_char().isalnum()):
                    self.tokens.append(Token(TokenType.WILDCARD, '_', self.line, self.column))
                    self.advance()
                    continue
            
            # Handle identifiers and keywords
            if self.current_char().isalpha() or self.current_char() == '_':
                self.tokens.append(self.tokenize_identifier_or_keyword())
                continue
            
            # Unknown character
            raise LexerError(f"Unexpected character: {self.current_char()}", 
                           self.line, self.column)
        
        # Add final DEDENTs to close all indentation levels
        while len(self.indent_stack) > 1:
            self.indent_stack.pop()
            self.tokens.append(Token(TokenType.DEDENT, 0, self.line, self.column))
        
        # Add EOF token
        self.tokens.append(Token(TokenType.EOF, None, self.line, self.column))
        
        return self.tokens
    
    def __iter__(self) -> Iterator[Token]:
        """Allow iteration over tokens."""
        if not self.tokens:
            self.tokenize()
        return iter(self.tokens) 

File: _legacy/src/bootstrap/src/lexer.rs
//! The Runa lexer, responsible for turning source code into a stream of tokens.
//! This is a production-ready implementation that correctly handles multi-word
//! keywords, indentation, and precise source location tracking.

use runa_common::token::{Token, TokenType};
use phf::phf_map;

// Single-word keywords - case-insensitive lookup
static KEYWORDS: phf::Map<&'static str, TokenType> = phf_map! {
    // Variable declarations
    "let" => TokenType::Let,
    "define" => TokenType::Define,
    "set" => TokenType::Set,
    "constant" => TokenType::Constant,
    
    // Control flow
    "if" => TokenType::If,
    "otherwise" => TokenType::Otherwise,
    "unless" => TokenType::Unless,
    "when" => TokenType::When,
    "match" => TokenType::Match,
    "switch" => TokenType::Switch,
    "case" => TokenType::Case,
    "default" => TokenType::Default,
    
    // Loops
    "for" => TokenType::For,
    "each" => TokenType::Each,
    "in" => TokenType::In,
    "while" => TokenType::While,
    "do" => TokenType::Do,
    "repeat" => TokenType::Repeat,
    "times" => TokenType::Times,
    "loop" => TokenType::Loop,
    "forever" => TokenType::Forever,
    "from" => TokenType::From,
    "to" => TokenType::To,
    "step" => TokenType::Step,
    "by" => TokenType::By,
    
    // Functions/Processes
    "process" => TokenType::Process,
    "called" => TokenType::Called,
    "that" => TokenType::That,
    "takes" => TokenType::Takes,
    "returns" => TokenType::Returns,
    "return" => TokenType::Return,
    "yield" => TokenType::Yield,
    "display" => TokenType::Display,
    
    // Connecting words
    "be" => TokenType::Be,
    "as" => TokenType::As,
    "with" => TokenType::With,
    "and" => TokenType::And,
    "or" => TokenType::Or,
    "not" => TokenType::Not,
    "of" => TokenType::Of,
    "the" => TokenType::The,
    "a" => TokenType::A,
    "an" => TokenType::An,
    "else" => TokenType::Else,

    // Indexing and Access
    "get" => TokenType::Get,
    "item" => TokenType::Item,
    "first" => TokenType::First,
    "last" => TokenType::Last,
    "key" => TokenType::Key,
    "value" => TokenType::Value,
    "index" => TokenType::Index,
    "at" => TokenType::AtWord,
    
    // Comments
    "note" => TokenType::Note,
    "end" => TokenType::End,
    
    // Types and Data Structures
    "type" => TokenType::Type,
    "struct" => TokenType::Struct,
    "enum" => TokenType::Enum,
    "list" => TokenType::List,
    "containing" => TokenType::Containing,
    "dictionary" => TokenType::Dictionary,
    "is" => TokenType::Is,
    "are" => TokenType::Are,
    
    // Boolean literals
    "true" => TokenType::Boolean,
    "false" => TokenType::Boolean,
    
    // Single-word operators
    "plus" => TokenType::Plus,
    "minus" => TokenType::Minus,
    "modulo" => TokenType::Modulo,
    
    // Function call keywords
    "call" => TokenType::Call,
    "invoke" => TokenType::Invoke,
    "execute" => TokenType::Execute,
};

// Multi-word operators and keywords - ordered from longest to shortest for maximal munch
static MULTI_WORD_OPERATORS: &[(&str, TokenType)] = &[
    // Multi-word comparison operators (longest first)
    ("is greater than or equal to", TokenType::IsGreaterThanOrEqualTo),
    ("is less than or equal to", TokenType::IsLessThanOrEqualTo),
    ("is not equal to", TokenType::IsNotEqualTo),
    ("is greater than", TokenType::IsGreaterThan),
    ("is less than", TokenType::IsLessThan),
    ("is equal to", TokenType::IsEqualTo),
    ("is not", TokenType::IsNot),  // Natural negation: "is not active"
    
    // Multi-word mathematical operators
    ("multiplied by", TokenType::MultipliedBy),
    ("divided by", TokenType::DividedBy),
    ("to the power of", TokenType::ToThePowerOf),
    
    // String concatenation operators
    ("followed by", TokenType::FollowedBy),
    ("concatenated with", TokenType::FollowedBy),
    ("joined with", TokenType::FollowedBy),

    
    // Function call syntax
    ("called with", TokenType::CalledWith),
];

pub struct Lexer<'a> {
    source: &'a str,
    chars: std::str::Chars<'a>,
    tokens: Vec<Token>,
    
    start_pos: usize,
    current_pos: usize,
    
    line: usize,
    column: usize,
    
    indent_stack: Vec<usize>,
    at_line_start: bool,
}

impl<'a> Lexer<'a> {
    pub fn new(source: &'a str) -> Self {
        Lexer {
            source,
            chars: source.chars(),
            tokens: Vec::new(),
            start_pos: 0,
            current_pos: 0,
            line: 1,
            column: 1,
            indent_stack: vec![0], // Base indentation is 0
            at_line_start: true,
        }
    }

    pub fn scan_tokens(&mut self) -> Result<Vec<Token>, String> {
        while let Some(_) = self.peek() {
            // Handle indentation at the start of each line
            if self.at_line_start {
                self.handle_indentation()?;
                self.at_line_start = false;
            }

            // Re-peek after indentation handling to get the correct character
            let Some(c) = self.peek() else { break; };
            self.start_pos = self.current_pos;
            
            match c {
                // Single-character tokens
                '(' => { self.advance(); self.add_token(TokenType::LParen); }
                ')' => { self.advance(); self.add_token(TokenType::RParen); }
                '{' => { self.advance(); self.add_token(TokenType::LBrace); }
                '}' => { self.advance(); self.add_token(TokenType::RBrace); }
                '[' => { self.advance(); self.add_token(TokenType::LBracket); }
                ']' => { self.advance(); self.add_token(TokenType::RBracket); }
                ',' => { self.advance(); self.add_token(TokenType::Comma); }
                '.' => { self.advance(); self.add_token(TokenType::Dot); }
                ':' => { self.advance(); self.add_token(TokenType::Colon); }
                ';' => { self.advance(); self.add_token(TokenType::Semicolon); }
                '|' => { self.advance(); self.add_token(TokenType::Pipe); }
                '?' => { self.advance(); self.add_token(TokenType::Question); }
                '!' => { self.advance(); self.add_token(TokenType::Exclamation); }
                '%' => { self.advance(); self.add_token(TokenType::Modulo); }
                '^' => { self.advance(); self.add_token(TokenType::ToThePowerOf); }
                
                // Assignment and arithmetic operators
                '=' => { self.advance(); self.add_token(TokenType::Assign); }
                '+' => { self.advance(); self.add_token(TokenType::Plus); }
                '-' => { self.advance(); self.add_token(TokenType::Minus); }
                '*' => { self.advance(); self.add_token(TokenType::MultipliedBy); }
                '/' => { self.advance(); self.add_token(TokenType::DividedBy); }
                
                // Whitespace
                ' ' | '\t' | '\r' => {
                    self.advance(); // Consume whitespace but don't create tokens
                }

                // Newlines
                '\n' => {
                    self.advance();
                    self.add_token(TokenType::Newline);
                    self.line += 1;
                    self.column = 1;
                    self.at_line_start = true;
                }

                // Strings
                '"' | '\'' => self.string()?,
                
                // Numbers
                c if c.is_ascii_digit() => self.number()?,
                
                // Raw string detection: special case for 'r' prefix
                'r' => {
                    // Look ahead for a quote by checking the source string directly
                    let next_pos = self.current_pos + 1;
                    if next_pos < self.source.len() {
                        let next_char = self.source.chars().nth(next_pos);
                        if next_char == Some('"') || next_char == Some('\'') {
                            self.advance(); // consume 'r'
                            self.raw_string()?; // handles the quote and content
                            continue;
                        }
                    }
                    self.identifier_or_keyword()?;
                },
                
                // Identifiers, keywords, and multi-word operators
                c if c.is_ascii_alphabetic() || c == '_' => self.identifier_or_keyword()?,
                
                // Annotations
                '@' => self.annotation()?,

                // Unexpected characters
                _ => {
                    return Err(format!("Unexpected character '{}' at line {} column {}", c, self.line, self.column));
                }
            }
        }
        
        // Emit any necessary Dedent tokens at the end
        while self.indent_stack.len() > 1 {
            self.indent_stack.pop();
            self.add_token_with_literal(TokenType::Dedent, "");
        }

        self.add_token(TokenType::Eof);
        Ok(self.tokens.clone())
    }

    /// Handle indentation at the start of a line
    fn handle_indentation(&mut self) -> Result<(), String> {
        let mut indent_level = 0;
        
        while let Some(c) = self.peek() {
            match c {
                ' ' => {
                    self.advance();
                    indent_level += 1;
                }
                '\t' => {
                    self.advance();
                    // Align tabs to the next multiple of 4
                    indent_level = (indent_level / 4 + 1) * 4;
                }
                _ => break, // Stop on any non-whitespace character
            }
        }

        // Ignore empty lines for indentation purposes
        if self.peek() == Some('\n') {
            return Ok(());
        }

        let current_indent = *self.indent_stack.last().unwrap();

        if indent_level > current_indent {
            self.indent_stack.push(indent_level);
            self.add_token_with_literal(TokenType::Indent, "");
        } else if indent_level < current_indent {
            while *self.indent_stack.last().unwrap() > indent_level {
                self.indent_stack.pop();
                self.add_token_with_literal(TokenType::Dedent, "");
            }
            // Check for inconsistent indentation
            if *self.indent_stack.last().unwrap() != indent_level {
                return Err(format!("Inconsistent indentation at line {}. Expected {} spaces, got {}", 
                    self.line, *self.indent_stack.last().unwrap(), indent_level));
            }
        }
        
        Ok(())
    }
    
    /// Handle identifiers, keywords, and multi-word operators with maximal munch
    fn identifier_or_keyword(&mut self) -> Result<(), String> {
        // First, try to match the longest possible multi-word operator
        for (op, token_type) in MULTI_WORD_OPERATORS {
            if self.source[self.current_pos..].starts_with(op) {
                // Check if this is a complete word boundary match
                let end_pos = self.current_pos + op.len();
                let is_word_boundary = end_pos >= self.source.len() || 
                    !self.source.chars().nth(end_pos).unwrap_or(' ').is_ascii_alphanumeric();
                
                if is_word_boundary {
                    // Found a match - advance past it and create the token
                    for _ in 0..op.len() {
                        self.advance();
                    }
                    self.add_token(*token_type);
                    return Ok(());
                }
            }
        }

        // If no multi-word operator matched, it's a single-word identifier or keyword
        self.advance_while(|c| c.is_ascii_alphanumeric() || c == '_');
        
        let lexeme = &self.source[self.start_pos..self.current_pos];
        
        // Look up in the single-word keyword map (case-insensitive)
        let token_type = KEYWORDS.get(lexeme.to_lowercase().as_str())
            .copied()
            .unwrap_or(TokenType::Identifier);
            
        self.add_token(token_type);
        Ok(())
    }
    
    /// Handle string literals with proper escape sequence support
    fn string(&mut self) -> Result<(), String> {
        let quote_char = self.advance().unwrap(); // Consume the opening quote
        let mut has_interpolation = false;
        
        while let Some(c) = self.peek() {
            if c == quote_char {
                break;
            }
            if c == '\\' {
                // Handle escape sequence
                self.advance(); // Consume '\'
                match self.peek() {
                    Some('n') | Some('t') | Some('r') | Some('\\') | 
                    Some('"') | Some('\'') => {
                        self.advance(); // Consume the escape character
                    }
                    Some('u') => {
                        self.advance(); // Consume 'u'
                        if self.peek() != Some('{') {
                            return Err("Invalid Unicode escape sequence".to_string());
                        }
                        self.advance(); // Consume '{'
                        // Parse hex digits
                        let mut hex_digits = String::new();
                        let mut found_digits = false;
                        while let Some(c) = self.peek() {
                            if c.is_ascii_hexdigit() {
                                self.advance();
                                hex_digits.push(c);
                                found_digits = true;
                            } else if c == '}' {
                                self.advance(); // Consume '}'
                                if !found_digits {
                                    return Err("Empty Unicode escape sequence".to_string());
                                }
                                break;
                            } else {
                                return Err("Invalid Unicode escape sequence".to_string());
                            }
                        }
                        if !found_digits {
                            return Err("Empty Unicode escape sequence".to_string());
                        }
                    }
                    _ => {
                        return Err("Invalid escape sequence".to_string());
                    }
                }
            } else {
                if c == '{' {
                    has_interpolation = true;
                }
                if c == '\n' {
                    self.line += 1;
                    self.column = 1;
                }
                self.advance();
            }
        }

        if self.peek().is_none() {
            return Err("Unterminated string".to_string());
        }

        self.advance(); // Consume the closing quote
        
        // Extract the string content (without quotes)
        let lexeme = &self.source[self.start_pos..self.current_pos];
        let token_type = if has_interpolation {
            TokenType::InterpolatedString
        } else {
            TokenType::String
        };
        self.add_token_with_literal(token_type, lexeme);
        Ok(())
    }

    /// Handle raw string literals (r"..." or r'...') - no escape sequence processing
    fn raw_string(&mut self) -> Result<(), String> {
        let quote_char = self.advance().unwrap(); // consume opening quote

        while let Some(c) = self.peek() {
            if c == quote_char {
                // Check if this quote is escaped (preceded by backslash)
                let prev_pos = self.current_pos.saturating_sub(1);
                if prev_pos >= self.start_pos {
                    let prev_char = self.source.chars().nth(prev_pos);
                    if prev_char == Some('\\') {
                        // This is an escaped quote, treat as literal
                        self.advance();
                        continue;
                    }
                }
                // This is an unescaped closing quote
                break;
            }
            if c == '\n' {
                self.line += 1;
                self.column = 1;
            }
            self.advance();
        }

        if self.peek().is_none() {
            return Err("Unterminated raw string".to_string());
        }

        self.advance(); // consume closing quote

        let lexeme = &self.source[self.start_pos..self.current_pos];
        self.add_token_with_literal(TokenType::String, lexeme);
        Ok(())
    }

    /// Handle numeric literals (integers and floats)
    fn number(&mut self) -> Result<(), String> {
        // Consume the first digit
        self.advance();
        
        // Consume remaining digits
        while let Some(c) = self.peek() {
            if c.is_ascii_digit() || c == '_' {
                self.advance();
            } else {
                break;
            }
        }

        // Look for decimal point
        if self.peek() == Some('.') {
            self.advance(); // Consume the '.'
            
            // Must have at least one digit after decimal point
            if !self.peek().map_or(false, |c| c.is_ascii_digit()) {
                return Err("Invalid float literal".to_string());
            }
            
            // Consume fractional digits
            while let Some(c) = self.peek() {
                if c.is_ascii_digit() || c == '_' {
                    self.advance();
                } else {
                    break;
                }
            }
            
            self.add_token(TokenType::Float);
        } else {
            self.add_token(TokenType::Integer);
        }
        
        Ok(())
    }

    /// Handle annotation tokens (e.g., @Reasoning, @Implementation)
    fn annotation(&mut self) -> Result<(), String> {
        self.advance(); // Consume the '@'
        
        let mut annotation_text = String::new();
        
        while let Some(c) = self.peek() {
            if c.is_ascii_alphabetic() || c == '_' {
                self.advance();
                annotation_text.push(c);
            } else {
                break;
            }
        }
        
        // Map annotation text to token type
        let token_type = match annotation_text.as_str() {
            "Reasoning" => TokenType::AtReasoning,
            "End_Reasoning" => TokenType::AtEndReasoning,
            "Implementation" => TokenType::AtImplementation,
            "End_Implementation" => TokenType::AtEndImplementation,
            "Uncertainty" => TokenType::AtUncertainty,
            "End_Uncertainty" => TokenType::AtEndUncertainty,
            "Request_Clarification" => TokenType::AtRequestClarification,
            "End_Request_Clarification" => TokenType::AtEndRequestClarification,
            "KnowledgeReference" => TokenType::AtKnowledgeReference,
            "End_KnowledgeReference" => TokenType::AtEndKnowledgeReference,
            "TestCases" => TokenType::AtTestCases,
            "End_TestCases" => TokenType::AtEndTestCases,
            "Resource_Constraints" => TokenType::AtResourceConstraints,
            "End_Resource_Constraints" => TokenType::AtEndResourceConstraints,
            "Security_Scope" => TokenType::AtSecurityScope,
            "End_Security_Scope" => TokenType::AtEndSecurityScope,
            "Execution_Model" => TokenType::AtExecutionModel,
            "End_Execution_Model" => TokenType::AtEndExecutionModel,
            "Performance_Hints" => TokenType::AtPerformanceHints,
            "End_Performance_Hints" => TokenType::AtEndPerformanceHints,
            "Progress" => TokenType::AtProgress,
            "End_Progress" => TokenType::AtEndProgress,
            "Feedback" => TokenType::AtFeedback,
            "End_Feedback" => TokenType::AtEndFeedback,
            "Translation_Note" => TokenType::AtTranslationNote,
            "End_Translation_Note" => TokenType::AtEndTranslationNote,
            "Error_Handling" => TokenType::AtErrorHandling,
            "End_Error_Handling" => TokenType::AtEndErrorHandling,
            "Request" => TokenType::AtRequest,
            "End_Request" => TokenType::AtEndRequest,
            "Context" => TokenType::AtContext,
            "End_Context" => TokenType::AtEndContext,
            "Task" => TokenType::AtTask,
            "End_Task" => TokenType::AtEndTask,
            "Requirements" => TokenType::AtRequirements,
            "End_Requirements" => TokenType::AtEndRequirements,
            "Verify" => TokenType::AtVerify,
            "End_Verify" => TokenType::AtEndVerify,
            "Collaboration" => TokenType::AtCollaboration,
            "End_Collaboration" => TokenType::AtEndCollaboration,
            "Iteration" => TokenType::AtIteration,
            "End_Iteration" => TokenType::AtEndIteration,
            "Clarification" => TokenType::AtClarification,
            "End_Clarification" => TokenType::AtEndClarification,
            _ => {
                return Err(format!("Unknown annotation: @{}", annotation_text));
            }
        };
        
        self.add_token_with_literal(token_type, &format!("@{}", annotation_text));
        Ok(())
    }

    // Core lexer primitives

    fn advance(&mut self) -> Option<char> {
        let char = self.chars.next()?;
        // For ASCII characters, len_utf8() should be 1, but let's be explicit
        self.current_pos += 1; // Always advance by 1 character position
        if char != '\n' && char != '\r' && char != '\t' {
            self.column += 1;
        }
        Some(char)
    }
    
    fn advance_while<F>(&mut self, mut predicate: F) 
    where 
        F: FnMut(char) -> bool 
    {
        while let Some(c) = self.peek() {
            if predicate(c) {
                self.advance();
            } else {
                break;
            }
        }
    }

    /// Peek at the next character without advancing
    fn peek_next(&self) -> Option<char> {
        let mut chars_copy = self.chars.clone();
        chars_copy.next(); // Skip current character
        chars_copy.next()  // Get next character
    }

    /// Peek at the current character without advancing
    fn peek(&self) -> Option<char> {
        self.chars.clone().next()
    }
    
    fn add_token(&mut self, token_type: TokenType) {
        let lexeme = &self.source[self.start_pos..self.current_pos];
        self.add_token_with_literal(token_type, lexeme);
    }

    fn add_token_with_literal(&mut self, token_type: TokenType, lexeme: &str) {
        let token_column = self.column.saturating_sub(lexeme.len());
        self.tokens.push(Token {
            token_type,
            lexeme: lexeme.to_string(),
            line: self.line,
            column: token_column,
        });
    }
} 